ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,295 @@
+
+
+#### CFFI
+
+[source](https://github.com/python-cffi/cffi) | [documentation](https://cffi.readthedocs.io/en/stable/)
+
+Licence: MIT
+Maturity: Start in 2012, many releases (33 tags on github), good documentation 
+Longevity: 8 maintainers (pypy), 77 contributors (github)
+Activity: Multiple releases in the last years, community interaction
+Visibility: 204 stars on Github 
+
+CFFI (C Foreign Function Interface) is a Python library that allows Python code
+to call C functions and use C data types. With CFFI it is possible to
+ 
+ * declare C interfaces: write C declarations (function prototypes, structs,
+   etc.) in a string using ffi.cdef()
+ * load C libraries (like .so, .dll, .dylib) using ffi.dlopen() and access all
+   symbols. 
+ * call C functions: python code can now call C functions and manipulate C data
+   structures as if they were native Python objects
+
+Using CFFI usally requires to write a certain amount of wrapping code to hide
+foreign structures and to make the usage of the library more pythonic. Exposing
+complex mechanisms (like an C++ iterator) requires writing and remapping
+intermediate C structures. The wrapping python code usually has to explicitly
+care about lifetime of objects. Python itself has no concept of ownership as it
+is considered in C/C++.
+
+
+#### PyBind11 3.0
+
+[source](https://github.com/pybind/pybind11) | [documentation](https://pybind11.readthedocs.io/en/stable/)
+
+Licence: BSD-style
+Maturity: Built on C++11 features, first released Oct 2015, 36 releases
+Longevity: 388 contributors (github)
+Activity: recent major release 3.0
+Visibility: 17.3k stars on Github, 26.6k users, used in PyTorch and TensorFlow
+
+pybind11 is a C++ header-only library that 
+enables seamless binding of C++ code to Python. It allows you to expose C++ classes, 
+functions, and data structures to Python with minimal boilerplate. 
+The python functionaly is directly expressed in C++ by
+ 
+ * writing a C++ wrapper using pybind11 macros (like PYBIND11\_MODULE).
+ * compile the C++ code into a Python extension module (usually done via the
+   buildsystem, e.g. CMake)
+ * import and use the C++ code directly from Python
+
+PyBind11 offers compatibility with most of C++ `std` (iterators, exceptions,
+smart pointers, variant/tuples, ...) and commonly used modules like NumPy.
+
+
+#### nanobind
+
+[source](https://github.com/wjakob/nanobind) | [documentation](https://nanobind.readthedocs.io/en/latest/)
+
+Licence: BSD-style
+Maturity: new project, start in 2022
+Longevity: 96 contributors (github)
+Activity: very new but high activity (44 Tags)
+Visibility: 3k stars on Github, 485k users
+
+nanobind is similar to PyBind11 (almost the same interface) but with the purpose of
+  
+  * being faster & having less overhead than pybind11
+  * producing smaller binary sizes
+
+while explicitly having reduced compatibility with `std` and third party modules.
+It is not header-only and requires linking to a shared or static library.
```

Comment:
Has a smaller range of functionality than pybind11. Note if anything specifically important.